Add Codespell to pre-commit

emolter opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Relates to [JP-3769](https://jira.stsci.edu/browse/JP-3769)

<!-- describe the changes comprising this PR here -->
This PR adds the [Codespell](https://github.com/codespell-project/codespell) spell-checker to the pre-commit workflow, and fixes all spelling issues in the repository.  This is part of a larger effort to add more code and documentation style rules to the repository; see https://github.com/spacetelescope/jwst/pull/9081

Thanks to @braingram for the suggestion to include Codespell.

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [ ] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

braingram commented on jwst/refpix/optimized_convolution.py:
```diff
@@ -167,11 +167,11 @@ def apply_conv_kernel(data, kernels, sigreject=4.0):
         normR = np.convolve(np.ones(nR.shape), kernel_r, mode='same')
         normR /= np.convolve(nR / 4, kernel_r, mode='same')
-        template = np.convolve(L, kernel_l, mode='same') * normL
+        template = np.convolve(L, kernel_l, mode='same') * norm_L
```
Why update `normL` and not `normR`? I vote for both or neither.

braingram commented on jwst/photom/miri_mrs.py:
```diff
@@ -144,7 +144,7 @@ def time_correction(input, detector, ftab, mid_time):
         # check that table 1 wavelength bin is an array with values
 
-    # Pull out the time coefficents for the detector we are working on
+    # Pull out the timecoefficients for the detector we are working on
```
```suggestion
    # Pull out the time coefficients for the detector we are working on
```

braingram commented on jwst/photom/miri_mrs.py:
```diff
@@ -64,7 +64,7 @@ def get_correction_function(side, timecoeff, mid_time):
         for every MRS band.
     timecoeff : dictionary
-        A dictionay holding the correction factors for the time/wavelength
+        Adictionary holding the correction factors for the time/wavelength
```
```suggestion
        A dictionary holding the correction factors for the time/wavelength
```

braingram commented on jwst/cube_build/src/cube_match_sky_pointcloud.c:
```diff
@@ -513,7 +513,7 @@ static PyObject *cube_wrapper(PyObject *module, PyObject *args) {
       (!(wave = ensure_array(waveo, &free_wave))) ||
       (!(flux = ensure_array(fluxo, &free_flux))) ||
-      (!(err = ensure_array(erro, &free_err))) ||
+      (!(err = ensure_array(err_o, &free_err))) ||
```
Similar to the other comments `erro` fits with `fluxo`, `waveo`, etc. I vote for updating all or none.

braingram commented on jwst/associations/lib/rules_level2_base.py:
```diff
@@ -558,7 +558,7 @@ def sort_by_candidate(asns):
         Notes
         -----
-        The current definition of candidates allows strictly lexigraphical
+        The current definition of candidates allows strictly lexicographical
```
```suggestion
        The current definition of candidates allows strictly alphabetical
```

braingram commented on jwst/associations/lib/product_utils.py:
```diff
@@ -21,7 +21,7 @@ def sort_by_candidate(asns):
     Notes
     -----
-    The current definition of candidates allows strictly lexigraphical
+    The current definition of candidates allows strictly lexicographical
```
```suggestion
    The current definition of candidates allows strictly alphabetical
```

braingram reviewed (commented):
Few comments/questions related to a few places where:
- spaces seemed to disappear, was this automatic?
- changes impacted variable names in a way that made them not match other variables

emolter commented:
Brett, sorry about the spaces disappearing - I fixed it locally and forgot to push.  It was caused by accidentally prepending whitespace during find & replace, and was not related to codespell itself.

RE the variable names, good points, I will fix those.  The reason I changed only some of them was because those were the only ones that codespell complained about - but I agree we should be consistent

braingram commented:
> Brett, sorry about the spaces disappearing - I fixed it locally and forgot to push. It was caused by accidentally prepending whitespace during find & replace, and was not related to codespell itself.
> 
> RE the variable names, good points, I will fix those. The reason I changed only some of them was because those were the only ones that codespell complained about - but I agree we should be consistent

Thanks! Any reasons to not enable `--write-changes` to have codespell make the changes automatically? One I can think of is it reduces the likelihood the developer will look at the changes but hopefully issues will be caught during review.

emolter commented:
> Any reasons to not enable --write-changes to have codespell make the changes automatically?

My reasoning was exactly that - the developer would be less likely to look at it themselves.

emolter commented:
With regards to adding `ruff`, `numpydoc`, and `codespell` to the test environment, ok, fine, they're not strictly used for the tests, but they're becoming a more important part of the development workflow.  So I kind of like the idea that someone should be able to easily install them all at once when setting up `jwst` for development.  Do you have an alternative suggestion?

braingram commented:
> My reasoning was exactly that - the developer would be less likely to look at it themselves.

Makes sense. My vote is to enable it for a few reasons:
- reduces the chance of find and replace errors
- (hopefully) reduces the number of fights with pre-commit
- ruff is already fixing code so this makes codespell consistent

My typical workflow with precommit is:
- make code codes
- stage commit changes
- run pre-commit
- when it fails it applies autofixes which aren't staged so a `git diff` shows all the changes (and the commit is prevented encouraging the developer to review the changes)

> With regards to adding ruff, numpydoc, and codespell to the test environment, ok, fine, they're not strictly used for the tests, but they're becoming a more important part of the development workflow. So I kind of like the idea that someone should be able to easily install them all at once when setting up jwst for development. Do you have an alternative suggestion?

I typically run `pip install pre-commit && pre-commit install`. When pre-commit runs it will check the configuration file and set up an environment with the pinned versions (all independent of the test environment). Even if these linting tools are part of the development environment they are not needed for something like `stasis` or the regression tests when it wants to run the tests. Adding them as `test` requirements complicates the test environment building (etc). I vote for leaving them off the `test` requirements. Another reason is that pre-commit likes to manage it's own dependency version pins (listed separately in the pre-commit config) and keeping these in sync with any versions we list in pyproject.toml would add to maintenance overhead.

tapastro commented:
I agree that they should not be part of the test environment - we use the test flag to build the SDP environment used in operations, and this would be unnecessary clutter for that environment. How about a "dev" dependency group that uses the "test" tag but also includes these style packages?

emolter commented:
ok, thanks both.  I hadn't realized that `pre-commit install` itself installs those, so in that case I think it's fine to just leave them out

braingram commented:
> dev = [
>     "ruff",
>     "numpydoc",
>     "codespel",
>     "jwst[test]"
> ]

Was the spelling of codespell intentional? 🤣

emolter commented:
Sounds good.  Just to keep the two PRs cleanly separated, with this one relating to Codespell only, I put the `dev` tag in the other one: https://github.com/spacetelescope/jwst/pull/9081/commits/3389375bf302a1e6c29f422dbb5f11aa2dd386a0

Does this look good to you @braingram @tapastro ?  I just tested it on a new conda env and it seems to work.  If so, I will update `CONTRIBUTING.md` about best practices for setting up our development environment, then get that PR, and finally update the config for this one and try to get it merged soon too.

And that sounds fine RE turning on auto-fixes

braingram commented:
> Sounds good. Just to keep the two PRs cleanly separated, with this one relating to Codespell only, I put the `dev` tag in the other one: [3389375](https://github.com/spacetelescope/jwst/commit/3389375bf302a1e6c29f422dbb5f11aa2dd386a0)
> 
> Does this look good to you @braingram @tapastro ? I just tested it on a new conda env and it seems to work. If so, I will update `CONTRIBUTING.md` about best practices for setting up our development environment, then get that PR, and finally update the config for this one and try to get it merged soon too.
> 
> And that sounds fine RE turning on auto-fixes

My vote is to have pre-commit manage the ruff, codespell, etc versions and installs. That's how things work in the CI and maintaining separate versions may lead to complications. Let's say someone right now installs the `dev` dependencies.  They will get ruff 0.9.3 whereas pre-commit lists 0.5.2. These will report and fix differ rule sets.

emolter commented:
> My vote is to have pre-commit manage the ruff, codespell, etc versions and installs. That's how things work in the CI and maintaining separate versions may lead to complications. Let's say someone right now installs the dev dependencies. They will get ruff 0.9.3 whereas pre-commit lists 0.5.2. These will report and fix different rule sets.

Melanie made the point that not everyone uses (or likes to use) `pre-commit` and may prefer to run tools individually.  I agree with your point that keeping the versions in-sync is somewhat of a problem though

tapastro commented:
> > Sounds good. Just to keep the two PRs cleanly separated, with this one relating to Codespell only, I put the `dev` tag in the other one: [3389375](https://github.com/spacetelescope/jwst/commit/3389375bf302a1e6c29f422dbb5f11aa2dd386a0)
> > Does this look good to you @braingram @tapastro ? I just tested it on a new conda env and it seems to work. If so, I will update `CONTRIBUTING.md` about best practices for setting up our development environment, then get that PR, and finally update the config for this one and try to get it merged soon too.
> > And that sounds fine RE turning on auto-fixes
> 
> My vote is to have pre-commit manage the ruff, codespell, etc versions and installs. That's how things work in the CI and maintaining separate versions may lead to complications. Let's say someone right now installs the `dev` dependencies. They will get ruff 0.9.3 whereas pre-commit lists 0.5.2. These will report and fix different rule sets.

Using pre-commit is fine, but I don't think part of our workflow should include requiring devs run `pip install pre-commit` - shouldn't our dependency files be able to handle that? Can we make a dev tag that installs pre-commit? I'll try to understand how pre-commit sets version requirements in the CI, but would we use a `.pre-commit-config.yaml` here?

emolter commented:
> I'll try to understand how pre-commit sets version requirements in the CI, but would we use a .pre-commit-config.yaml here?

We already do use `.pre-commit-config.yaml` and yes, that is where pre-commit sets version requirements.

One thing I don't really like about suggesting to only use `pip install pre-commit && pre-commit install` is that after those two commands, running (e.g.) `ruff check .` gives `-bash: ruff: command not found`.  It feels like whatever commands we suggest that devs use should have the end result that all of the tools will be installed and work individually as well as together.

It would be really nice if there was an easy/robust way to keep the versions in sync... but I think this goes counter to the philosophy of pre-commit based on [this stackoverflow answer from the creator of pre-commit](https://stackoverflow.com/a/72888197).

emolter commented:
Regression tests started here: https://github.com/spacetelescope/RegressionTests/actions/runs/13036860875 

Build failures are the same as on all PRs right now

braingram reviewed (approved):
LGTM. There's a merge conflict which looks minor. All regtests passed and I think given the scope of changes a maitainer review will also be required.

melanieclarke commented on jwst/ami/utils.py:
```diff
@@ -52,7 +52,7 @@ class Affine2d:
     and need to calculate G(u,v) at a grid of points in the (u,v) space, with
     two lattice vectors a and b defining the grid.  These lattice vectors have
-    components a=(au,av) and b=(bu,bv) along the u and v axes.
+    components a=(au,av) and b=(bu,bv) along the u and v axes. # codespell:ignore bu
```
This codespell comment appears in the docstring, which I don't love.  Is there some other way to get it to ignore this line? Or else rephrase the docstring?

melanieclarke replied:
Maybe like this?
```suggestion
    components a=(a_u,a_v) and b=(b_u,b_v) along the u and v axes.
```

emolter replied:
I just did exactly that :)

melanieclarke reviewed (commented):
This PR makes my inner copy-editor so happy.  One question/objection, but mostly this looks great.

melanieclarke commented:
I'm happy with this, but I'll hold off on approval in case Tyler has comments.

tapastro reviewed (approved):
Looks good! I'm impressed that it caught the covert -> convert fix.
